File: src/utils/activity_manager.rs
use poise::serenity_prelude::{ActivityData, ShardManager};
use rand::prelude::IndexedRandom;
use std::sync::{Arc, OnceLock};
use std::time::Duration;
use tokio::time::interval;
use tracing::info;

static ACTIVITIES: OnceLock<Vec<ActivityData>> = OnceLock::new();

fn get_activities() -> &'static Vec<ActivityData> {
  ACTIVITIES.get_or_init(|| {
    vec![
      ActivityData::playing("Pokémon"),
      ActivityData::playing("Gooning Aim Trainer"),
      ActivityData::playing("Battletoads"),
      ActivityData::playing("Counter-Strike 2"),
      ActivityData::playing("Hello Kitty Island Adventure"),
      ActivityData::playing("Badminton"),
      ActivityData::watching("JasonTheWeen"),
      ActivityData::watching("xQc"),
      ActivityData::watching("Pokimane"),
      ActivityData::watching("Dantes"),
      ActivityData::listening("ZWE1HVNDXR"),
      ActivityData::listening("Illenium"),
      ActivityData::listening("The Chainsmokers"),
      ActivityData::listening("KSI"),
      ActivityData::custom("Going to Plan B"),
      ActivityData::custom("Clubbing at Mission"),
      ActivityData::custom("Waiting in line at Den Social"),
    ]
  })
}

pub fn start_activity_rotation(shard_manager: Arc<ShardManager>) {
  tokio::spawn(async move {
    let mut interval = interval(Duration::from_secs(60 * 60));
    loop {
      interval.tick().await;

      let activity_data = get_activities().choose(&mut rand::rng()).unwrap().clone();

      info!("Setting activity data to\n{:#?}", activity_data);
      let runners = shard_manager.runners.lock().await;
      runners.iter().for_each(|(_, runner)| {
        info!("Setting activity data for runner:\n{:#?}", runner);
        runner.runner_tx.set_activity(Some(activity_data.clone()));
      });
    }
  });
}

File: src/utils/activity-manager.ts
import { Client, ActivityType } from 'discord.js'
import { logger } from './logger'

export const startActivityRotation = (client: Client) => {
  logger.info('Starting activity rotation...')
  const rotateActivity = () => {
    const activity = getRandomActivity()
    logger.info({ activity }, 'Setting activity data')
    client.user?.setActivity(activity.name, { type: activity.type })
  }

  rotateActivity()
  setInterval(rotateActivity, 60 * 60 * 1000)
  logger.info('Activity rotation started!')
}

interface ActivityData {
  type: ActivityType;
  name: string;
}

const ACTIVITIES: ActivityData[] = [
  { type: ActivityType.Playing, name: 'Pokémon' },
  { type: ActivityType.Playing, name: 'Gooning Aim Trainer' },
  { type: ActivityType.Playing, name: 'Battletoads' },
  { type: ActivityType.Playing, name: 'Counter-Strike 2' },
  { type: ActivityType.Playing, name: 'Hello Kitty Island Adventure' },
  { type: ActivityType.Playing, name: 'Badminton' },
  { type: ActivityType.Watching, name: 'JasonTheWeen' },
  { type: ActivityType.Watching, name: 'xQc' },
  { type: ActivityType.Watching, name: 'Pokimane' },
  { type: ActivityType.Watching, name: 'Dantes' },
  { type: ActivityType.Listening, name: 'ZWE1HVNDXR' },
  { type: ActivityType.Listening, name: 'Illenium' },
  { type: ActivityType.Listening, name: 'The Chainsmokers' },
  { type: ActivityType.Listening, name: 'KSI' },
  { type: ActivityType.Custom, name: 'Going to Plan B' },
  { type: ActivityType.Custom, name: 'Clubbing at Mission' },
  { type: ActivityType.Custom, name: 'Waiting in line at Den Social' },
]

const getRandomActivity = (): ActivityData => {
  const randomIndex = Math.floor(Math.random() * ACTIVITIES.length)
  return ACTIVITIES[randomIndex]
}
